src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -492,17 +450,21 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
         match *place {
             Place::Local(ref local) => self.visit_local(local, context, location),
             Place::Static(ref global) => {
-                self.add(Qualif::STATIC);
+                // Only allow statics (not consts) to refer to other statics.
+                if !(self.mode == Mode::Static || self.mode == Mode::StaticMut) {
+                    self.add(Qualif::NOT_CONST);
+                }
 
                 if self.mode != Mode::Fn {
-                    for attr in &self.tcx.get_attrs(global.def_id)[..] {
-                        if attr.check_name("thread_local") {
-                            span_err!(self.tcx.sess, self.span, E0625,
-                                      "thread-local statics cannot be \
-                                       accessed at compile-time");
-                            self.add(Qualif::NOT_CONST);
-                            return;
-                        }
+                    if self.tcx
+                           .get_attrs(global.def_id)
+                           .iter()
+                           .any(|attr| attr.check_name("thread_local")) {
+                        span_err!(self.tcx.sess, self.span, E0625,
+                                    "thread-local statics cannot be \
+                                    accessed at compile-time");
+                        self.add(Qualif::NOT_CONST);
+                        return;
```

Comment:
You can also changed this so the attribute is *always* checked, and `self.add(Qualif::NOT_CONST); return;` always happen, only the error reporting stays conditional on `if self.mode != Mode::Fn`.

Reply:
Makes sense.